src/functional/ffront/decorator.py

```diff
@@ -542,11 +547,11 @@ def as_program(
             location=loc,
         )
         out_ref = past.Name(id="out", location=loc)
-
+        assert isinstance(self.foast_node, (foast.FieldOperator, past.Program, foast.ScanOperator))
```

Comment:
I think the correct thing to do is to *constrain* `OperatorNodeT` on `foast.FieldOperator` and `foast.ScanOperator` because not other type is meaningful there currently. This solves all the problems, the only downside is `FieldOperator.from_function` can not have a default argument for `operator_node_cls`, as that argument is required by mypy to match up with whatever type `OperatorNodeT` represents in the concrete (non-generic) `FieldOperator[OperatorNodeT]` in question.

In practice this just means passing this argument in one place.